### PR TITLE
First PR as a project initialization

### PR DESCRIPTION
**Pull request by mizaqq**
PR contains:
- Basic CNN clasifier for leaves. 
- Dataset split into train-test sets
- Dataloader for concurrent data loading during training
- Basic model training

TODO:
-add hydra for better configs
-add mlflow for logging
-further model development

### CODE REVIEW

File: .vscode/launch.json [hunk 1]
```diff
@@ -0,0 +1,18 @@
+{
+    // Użyj funkcji IntelliSense, aby uzyskać informacje o możliwych atrybutach.
+    // Najedź kursorem, aby wyświetlić opisy istniejących atrybutów.
+    // Aby uzyskać więcej informacji, odwiedź stronę: https://go.microsoft.com/fwlink/?linkid=830387
```
Comment (adamwawrzynski) [inline .vscode/launch.json]:
```suggestion
```

File: .vscode/launch.json [hunk 2]
```diff
@@ -0,0 +1,18 @@
...
+    "version": "0.2.0",
+    "configurations": [
+        {
+            "name": "Debuger języka Python: Current File",
```
Comment (adamwawrzynski) [inline .vscode/launch.json]: Keep names in English.

File: src/main.py [hunk 3]
```diff
@@ -0,0 +1,18 @@
+import torch
+
+from src.model.cnn import Convolutional
+from src.model.train import Model
+from src.preprocessing.dataloader import Dataloader
+
+
+def main() -> None:
+    dataloader = Dataloader(workers=8)
+    device = torch.device("cuda" if torch.cuda.is_available() else "cpu")
```
Comment (adamwawrzynski) [inline src/main.py:10]: I would move selection of `workers` and `device` as arguments of the function + function to fetch those parameters from CLI.

File: src/model/cnn.py [hunk 4]
```diff
@@ -0,0 +1,35 @@
...
+
+    def forward(self, x: torch.Tensor) -> torch.Tensor:
+        x = self.pool1(F.relu(self.conv1(x)))
+        x = self.pool2(F.relu(self.conv2(x)))
+        x = self.pool3(F.relu(self.conv3(x)))
+        x = x.view(-1, self.num_flat_features(x))
+        x = F.relu(self.fc1(x))
+        x = self.fc2(x)
+        x = F.softmax(x, dim=1)
+        return x
```
Comment (adamwawrzynski) [inline src/model/cnn.py:28]: Try to rewrite this method to be more generic and less prone to change when `__init__` method will change. I mean you could use list of layers instead of directly using them.

File: src/model/cnn.py [hunk 5]
```diff
@@ -0,0 +1,35 @@
...
+        self.pool1 = torch.nn.MaxPool2d(2)
+        self.conv2 = torch.nn.Conv2d(32, 64, 3, 2, padding_mode='replicate')
+        self.pool2 = torch.nn.MaxPool2d(2)
+        self.conv3 = torch.nn.Conv2d(64, 128, 3, 2, padding_mode='replicate')
+        self.pool3 = torch.nn.MaxPool2d(2)
+        self.conv4 = torch.nn.Conv2d(128, 256, 3, 2, padding_mode='replicate')
+        self.pool4 = torch.nn.MaxPool2d(2)
+        self.conv5 = torch.nn.Conv2d(256, 512, 3, 2, padding_mode='replicate')
+        self.fc1 = torch.nn.Linear(1152, 32)
+        self.fc2 = torch.nn.Linear(32, 4)
```
Comment (adamwawrzynski) [inline src/model/cnn.py:18]: Try to figure out way of creating and parametrizing network architecture for ease of experimentation with different parameters.

File: src/model/cnn.py [hunk 6]
```diff
@@ -0,0 +1,35 @@
...
+        x = self.fc2(x)
+        x = F.softmax(x, dim=1)
+        return x
+
+    def num_flat_features(self, x: torch.Tensor) -> int:
+        size = x.size()[1:]
+        num_features = 1
+        for s in size:
+            num_features *= s
+        return num_features
```
Comment (adamwawrzynski) [inline src/model/cnn.py]: Use `@staticmethod` or make it normal function as you are not using `self` anywhere.

File: src/model/train.py [hunk 7]
```diff
@@ -0,0 +1,46 @@
...
+        total = 0
+        with torch.no_grad():
+            for data in self.data_loader.test_loader:
+                images, labels = data
+                images, labels = images.to(self.device), labels.to(self.device)
+                outputs = self.model(images)
+                _, predicted = torch.max(outputs.data, 1)
+                total += labels.size(0)
+                correct += (predicted == labels).sum().item()
+        print(f'Accuracy of the network on the test images: {100 * correct / total}%')
```
Comment (adamwawrzynski) [inline src/model/train.py]: Make it more generic. Re-write this method to obtain results in a form of ground truth labels and predicted labels for easier change in used metrics. I recommend using sklearn classification report function as a baseline for any classification tasks as it provides not only accuracy, recall, and f-1 score but also micro and macro scores. Also you might want to add confusion matrix computation.

In general parametrize the class with metrics you want to use in the constructor for easier change.

File: src/preprocessing/dataloader.py [hunk 8]
```diff
@@ -0,0 +1,48 @@
...
+        self.test_loader = self.get_loader(self.test_data, shuffle=False)
+
+    def get_transformer(self, size: int, crop: int) -> transforms.Compose:
+        return transforms.Compose(
+            [
+                transforms.Resize(size),
+                transforms.CenterCrop(crop),
+                transforms.ToTensor(),
+            ]
+        )
```
Comment (adamwawrzynski) [inline src/preprocessing/dataloader.py]: Mark it as `@staticmethod` as you are not using `self` argument.

File: src/model/train.py [hunk 9]
```diff
@@ -0,0 +1,46 @@
...
+class Model:
+    def __init__(self, model: torch.nn.Module, device: torch.device, data_loader: Dataloader) -> None:
+        self.model = model
+        self.device = device
+        self.data_loader = data_loader
+
+    def train_model(self, lr: float = 0.001, momentum: float = 0.9, epochs: int = 10) -> None:
+        self.model.to(self.device)
+        loss_fn = torch.nn.CrossEntropyLoss()
+        optimizer = torch.optim.SGD(self.model.parameters(), lr=lr, momentum=momentum)
```
Comment (adamwawrzynski) [inline src/model/train.py]: I would recommend having loss function ans optimizer passed as arguments in the constructor of this class. In general, anything you might want to experiment with should be parametrized, e.g. in the constructor of a class.

File: poetry.lock
Comment (adamwawrzynski) [inline poetry.lock:1]: General comment: use code from `src` directory/package instead of re-implementing functions and classes in the notebook. It will prevent from accidentally experimenting on different architecture than you have implemented in the `src` code which will be used for final training.

File: .vscode/launch.json (hunk 1 above)
Comment (mizaqq) [reply .vscode/launch.json]: [deleted](https://github.com/mizaqq/plant-disease/pull/1/commits/eb8238e62e85caf2190035ad447a14d2bdc98831)

File: .vscode/launch.json (hunk 2 above)
Comment (mizaqq) [reply .vscode/launch.json]: [deleted](https://github.com/mizaqq/plant-disease/pull/1/commits/eb8238e62e85caf2190035ad447a14d2bdc98831)

File: src/model/cnn.py (hunk 5 above)
Comment (mizaqq) [reply src/model/cnn.py:18]: Done in the following PR which utilizes [hydra](https://github.com/mizaqq/plant-disease/commit/3eaca35a329b34b61d3e692670490f70cd9d1c58)

File: src/model/cnn.py (hunk 6 above)
Comment (mizaqq) [reply src/model/cnn.py]: [done](https://github.com/mizaqq/plant-disease/pull/1/commits/63c0c34f70203b783f43ebeb52302b5a4757f1b5)

File: src/preprocessing/dataloader.py (hunk 8 above)
Comment (mizaqq) [reply src/preprocessing/dataloader.py]: [done](https://github.com/mizaqq/plant-disease/pull/1/commits/a8b24ecc3172c9b1cd058dddadc49a614930dba4)

File: src/model/train.py (hunk 9 above)
Comment (mizaqq) [reply src/model/train.py]: [done](https://github.com/mizaqq/plant-disease/pull/1/commits/fc39cf3f5c0d559db8df09329c95c568136c3807)

File: src/model/train.py (hunk 7 above)
Comment (mizaqq) [reply src/model/train.py]: [refactored](https://github.com/mizaqq/plant-disease/pull/1/commits/5c1417c6d6e62242b06ffece6432f3b11e66196d)

File: src/model/cnn.py (hunk 4 above)
Comment (mizaqq) [reply src/model/cnn.py:28]: [Done](https://github.com/mizaqq/plant-disease/commit/f785a1b12ef7541d2961be41cc4d294e825f3983) in the following PR
